openhtf/io/output/json_factory.py

```diff
@@ -51,12 +50,17 @@ def __call__(self, test_record):
     else:
       as_dict = data.ConvertToBaseTypes(test_record, ignore_keys='attachments')
     if isinstance(self.filename_pattern, basestring):
-      if '{' in self.filename_pattern: 
-        filename = self.filename_pattern.format(**as_dict)
+      if '{' in self.filename_pattern:
+        self.filename = self.filename_pattern.format(**as_dict)
       else:
-        filename = self.filename_pattern % as_dict
-      with open(filename, 'w') as f:
-        f.write(self.encode(as_dict))
+        self.filename = self.filename_pattern % as_dict
+      self.file_data = self.encode(as_dict)
+      self.save_file()
     else:
       self.filename_pattern.write(self.encode(as_dict))
   # pylint: enable=invalid-name
+
+  def save_file(self):
```

Comment:
there's still a lot of code duplication here.  ideally, the output module (`io/output/__init__.py`) would contain an `OutputToFile` base class, where the `.format()` and file-writing functionality would reside, and then subclasses would implement the test-record-to-file-data mechanism.

in fact, test records are now pickle-able, so the default file output mechanism should be to simply pickle the test record and write that out (a very pure-python way to go about it), but subclasses can override that method to do other output formats (making those subclasses very small and clean).